File: Algorithm_155_Min Stack.cpp
#include <iostream>
#include <vector>
using namespace std;
class MinStack {
private:
	vector<int> sta;
	int len;
public:
    /** initialize your data structure here. */
    MinStack() {
        len=0;
    }
    
    void push(int x) {
        sta.push_back(x);
        len++;
    }
    
    void pop() {
    	if (len==0) return;
        sta.erase(sta.begin()+len-1);
        len--;
    }
    
    int top() {
        if (len>=1) return sta[len-1];
    }
    
    int getMin() {
    	int min=sta[0];
        for (int i=1;i<len;i++) if (sta[i]<min) min=sta[i];
		return min; 
    }
};

File: Algorithm_155_Min Stack.py
class MinStack:

    def __init__(self):
        self.stack = []
        self.min = None

    def push(self, x: int) -> None:
        # x is the new min element
        # must use <= here, because one element may be pushed for many times
        # use self.min == None instead of not self.min, because not 0 == True
        if self.min == None or x <= self.min:
            self.stack.append(self.min)
            self.min = x
        self.stack.append(x)

    def pop(self) -> None:
        if not self.stack:
            return
        x = self.stack.pop()
        # x is the min element, update self.min
        if x == self.min:
            self.min = self.stack.pop()

    def top(self) -> int:
        return self.stack[-1]

    def getMin(self) -> int:
        return self.min

# Your MinStack object will be instantiated and called as such:
# obj = MinStack()
# obj.push(x)
# obj.pop()
# param_3 = obj.top()
# param_4 = obj.getMin()